File: scripts/limpar.js
// Consolida os arquivos JSON baixados em um único arquivo e remove dados irrelevantes.

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { stripHtml } from 'string-strip-html';
import { createHash } from 'node:crypto';

const __dirname = path.dirname(fileURLToPath(import.meta.url)) + '/';

const downloadsDir = path.join(__dirname, '../data/downloads');
const arquivosJson = fs
    .readdirSync(downloadsDir)
    .filter((file) => file.endsWith('.json'));

// programação das últimas 10 edições da Campus Party, separadas por edições
const programacao = [];

for (const arquivoJson of arquivosJson) {
    console.log(`Processando: "${arquivoJson}"`);

    const edicaoNome = /(?<=\d - ).*(?=\.json)/
        .exec(arquivoJson)[0]
        .replace('-bancadas', '');

    const json = JSON.parse(
        fs.readFileSync(path.join(downloadsDir, arquivoJson), 'utf8'),
    );

    json.result.schedule
        .map((s) => s.activity_list)
        .flat()
        .forEach((atividade) => {
            const atividadeLimpa = {
                name: atividade.name,
                auditorium: atividade.auditorium_name,
                description: stripHtml(atividade.description.trim()).result,
                open: atividade.open,
                start_time: atividade.start_time,
                end_time: atividade.end_time,
                speakers: atividade.speakers.map((speaker) => ({
                    name: speaker.name,
                    profession: speaker.profession,
                    description: stripHtml((speaker.description ?? '').trim())
                        .result,
                })),
            };

            const uid = gerarUid(atividadeLimpa);

            // descarta duplicatas
            if (programacao.some((item) => item.uid === uid)) return;

            programacao.push({
                uid,
                edicao: edicaoNome,
                ...atividadeLimpa,
            });
        });
}

escreverArquivo(path.join(__dirname, '../data/limpo.json'), programacao);

function escreverArquivo(caminho, data) {
    try {
        fs.writeFileSync(caminho, JSON.stringify(data, null, 2), 'utf8');
        console.log(`Arquivo escrito: ${caminho}`);
    } catch (error) {
        console.error('Error ao tentar escrever o arquivo:', error);
    }
}

export function gerarUid(obj) {
    const str = JSON.stringify(obj);
    const hash = createHash('sha256');
    hash.update(str, 'utf8');
    const reproducibleId = hash.digest('hex');
    return reproducibleId;
}

File: src/limpar.py
# Consolida os arquivos JSON baixados em um único arquivo e remove dados irrelevantes.

import os
import json
import pandas as pd
import hashlib
from bs4 import BeautifulSoup

src_dir = os.path.dirname(os.path.abspath(__file__))
data_dir = os.path.join(src_dir, '..', 'data')
downloads_dir = os.path.join(data_dir, 'downloads')

def gerar_uid(df):
    str_obj = df.drop(['uid', 'edicao']).to_json(force_ascii=False)
    return hashlib.sha256(str_obj.encode('utf-8')).hexdigest()

def strip(text):
    return BeautifulSoup(text, "html.parser").get_text().strip()

def main() :
    columns = ['uid', 'edicao', 'name', 'auditorium', 'description', 'open', 'start_time', 'end_time', 'speakers']
    programacao = pd.DataFrame(columns=columns)

    for arquivoJson in sorted(os.listdir(downloads_dir)):
        if not arquivoJson.endswith('.json'):
            continue

        print(f"Processando: {arquivoJson}")

        edicaoNome = arquivoJson.split(' - ')[1].replace('-bancadas', '').replace('.json', '')

        with open(os.path.join(downloads_dir, arquivoJson), 'r', encoding='utf-8') as f:
            json_data = json.load(f)

        for atividade in json_data['result']['schedule']:
            for item in atividade['activity_list']:
                atividadeLimpa = pd.Series({
                    'name': item['name'],
                    'auditorium': item['auditorium_name'],
                    'description': strip(item['description']),
                    'open': item['open'],
                    'start_time': item['start_time'],
                    'end_time': item['end_time'],
                    'speakers': [{'name': speaker['name'], 'profession': speaker['profession'], 'description': strip(speaker.get('description') or '')} for speaker in item['speakers']]
                }, index=columns)

                uid = gerar_uid(atividadeLimpa)

                atividadeLimpa['uid'] = uid
                atividadeLimpa['edicao'] = edicaoNome
                programacao.loc[len(programacao)] = atividadeLimpa

    programacao.drop_duplicates(subset=['uid'], inplace=True)

    programacao.to_json(os.path.join(data_dir, 'limpo.json'), orient='records', force_ascii=False, indent=2)

if __name__ == "__main__":
    main()
